updates to XRPC service auth spec (aud and iss fields) (#616)
bluesky-social/atproto-website: src/app/[locale]/specs/xrpc/en.mdx

    - use `ES256K` for `k256` keys
    - use `ES256` for `p256` keys
- `typ` header field (string, required): currently `JWT`, but intend to update to a more specific value.

> - `kid` header field (string, optional): issuer key identifier, used to specify which public key was used to sign the JWT. Corresponds to identifier fragment in DID document, specifying which "verification method" was used. This field should not include the DID or the hash separator (`#`). Defaults to `atproto`. Note that this field is in the (unsigned) JWT header, not the body. See below for semantics and verification guidance.

Reviewer: I would prefer to use the DID URL syntax here:

can be either:
 - relative (to issuer) `#my-key`
 - absolute `did:plc:123#my-key` (with a DID that MUST be equal to the issuer)

defaulting to `#atproto`.

If we don't want absolute DID-URLs, I would still keep the `#` prefix as it keeps the syntax more consistent with the DID spec. It also makes the syntax distinct from what is typically used in oauth land, which should help avoiding confusion as to where to find this key when encountering such tokens.

Reviewer: ```suggestion
- `kid` header field (string, optional): issuer key identifier, used to specify which public key was used to sign the JWT. Corresponds to identifier fragment in DID document, specifying which "verification method" was used. This field should not include the DID or the hash separator (`#`). Defaults to `atproto`. See below for semantics and verification guidance.
```

Author: I'm against having the `iss` DID repeated in this field. It feels redundant/inefficient, and implementation complexity (another thing to verify).

I don't have super strong feelings about whether to include the `#`, but would lean towards not? If feels unexpected.

I might be open to other syntax to differentiate this field from other `kid` types? But keeping it simple and just being the identifier value feels like the best move.

Open to discussing further.

Author: we decided to go ahead with `#` prefix in the `kid` field, but not the full DID.

- `iss` body field (string, required): account DID ("issuer") that the request is being sent on behalf of.

> - `aud` body field (string, required): service DID associated with the service that the request is being sent to ("audience"). May include an optional fragment name identifying the specific service type and endpoint. See [DID syntax specification](/specs/did) for syntax details, and below for semantics. New software should always include the service fragment name when generating JWTs.

Reviewer: ```suggestion
- `aud` body field (string, required): service DID associated with the service that the request is being sent to ("audience"). Should include a fragment name identifying the specific service type and endpoint. New software should always include the service fragment name when generating JWTs. Legacy services may accept `aud` values without fragment. See [DID syntax specification](/specs/did) for syntax details, and below for semantics.
```

- `exp` body field (number, required): token expiration time, as a UNIX timestamp with seconds precision. Should be a short time window, as revocation is not implemented. 60 seconds is a good token lifetime.
- `iat` body field (number, required): token creation time, as a UNIX timestamp with seconds precision.
- `lxm` body field (string, optional): short for "lexicon method". NSID syntax. Indicates the endpoint that this token authorizes. This field is required for all authenticated XRPC endpoint requests.

> - `jti` body field (string, required): unique random string nonce. May be used by receiving services to prevent reuse of token and replay attacks.

Reviewer: ```suggestion
- `jti` body field (string, required): unique random string nonce. Should be used by receiving services to prevent reuse of token and replay attacks.
```

- JWT signature (string, required): base64url-encoded signature using the account DID's signing key.

> The issuer field (`iss`) is the DID of the account sending the request. If an issuer key identifier (`kid`) field is provided, it indicates a specific public key from the account's DID document in the `verificationMethod` array, defaulting to `atproto`. Note that `iss` is a JWT body field, while `kid` is a JWT header field. Receiving services must only accept specific key types (verification method identifiers), to maintain security separation between registered keys. This usually means accepting only `atproto`, unless there is a specific application-specific or service-specific additional key type expected (eg, `atproto_label`).

Reviewer: Add something like:

> Recipients must always check that the key referenced by the `kid` (if present) matches a verification method identifier present in the DID Document of the issuer (`iss`).

Author: I tweaked the paragraph a bit to make this clearer.

The HTTP header is `atproto-proxy`, and the value is a DID (identifying a service), followed by a service endpoint identifier, joined by a `#` character. The PDS resolves the service DID, extracts a service endpoint URL from the DID document, and proxies the request on to the identified server.

<Note>
Until Spring 2026, the reference PDS implementation did not pass through the full service reference to the `aud` field of the JWT auth header when proxying requests. It would strip the service endpoint identifier, and only include the bare DID.

> This is in the process of being updated to include both the DID and service identifier fragment.

Reviewer 2: Should we implement this before merging so we aren't stuck with "in the process of being updated"? Otherwise this looks good.

Author: There is going to need to be a transition period between when we publish this update (and update our services/SDKs), and when everybody else in the ecosystem (eg, feed gens and appviews) do the updates on their end.

I'll lay this out in the (forthcoming) updated proposal.

Author: Here is that full proposal, including transition plan: https://github.com/bluesky-social/proposals/tree/main/0014-service-auth-revised

Reviewer 3: I think the "Until Spring 2026, the reference PDS implementation did not pass through" part is slightly misleading, because AFAICT it still doesn't pass through?…
